Refactor typesupport

jacobperron opened the pull request:
Looking for early feedback. I think the EmPy templates are complete, but there is still some CMake business to do.

dirk-thomas committed:
update includes and type mapping

jacobperron committed:
Update python, cmake, and msg resource files

jacobperron committed:
Messages working now

jacobperron committed:
Update service EmPy files

jacobperron committed:
cleanup

jacobperron committed:
[wip] typesupport for opensplice c refactor

dirk-thomas commented on rosidl_typesupport_opensplice_cpp/bin/rosidl_typesupport_opensplice_cpp:
```diff
@@ -40,7 +40,7 @@ def main(argv=sys.argv[1:]):
 
     try:
-        rc = generate_typesupport_opensplice_cpp(generator_args)
+        rc = generate_typesupport_opensplice_cpp(args.generator_arguments_file)
```
Line 40 is obsolete then.

jacobperron replied:
`generator_args` is used at L50.

dirk-thomas commented on rosidl_typesupport_opensplice_cpp/cmake/rosidl_typesupport_opensplice_cpp_generate_interfaces.cmake:
```diff
@@ -12,85 +12,79 @@
-    list(APPEND _dds_idl_files
-      "${_dds_idl_base_path}/${PROJECT_NAME}/${_parent_folder}/dds_opensplice/Sample_${_name}_Response_.idl")
-  endif()
-endforeach()
```
Have you tried if the OpenSplice code generator can handle the currently generated `.idl` files? Afaik the Connext one e.g. can't handle the `(u)intN` types in the current version.

dirk-thomas commented on rosidl_typesupport_opensplice_cpp/cmake/rosidl_typesupport_opensplice_cpp_generate_interfaces.cmake:
```diff
@@ -12,85 +12,79 @@
   get_filename_component(_parent_folder "${_parent_folder}" NAME)
-  get_filename_component(_extension "${_idl_file}" EXT)
-  get_filename_component(_msg_name "${_idl_file}" NAME_WE)
+  get_filename_component(_msg_name "${_rel_idl_file}" NAME_WE)
```
I suggest to rename `_msg_name` to `_idl_name` since this might not be a message.

dirk-thomas commented on rosidl_typesupport_opensplice_cpp/resource/msg__type_support.cpp.em:
```diff
@@ -122,55 +142,57 @@ register_type__@(spec.base_type.type)(
 @[    else]@
-    size_t size = ros_message.@(field.name).size();
+    size_t size = ros_message.@(member.name).size();
+@[    end if]@
```
The `else` block ends 5 lines earlier now but it looks like the following four lines are only necessary in the else case?

jacobperron replied:
I accidentally changed the logic. Doesn't seem like it hurts as it is now, but I can change it back just in case.

dirk-thomas commented on rosidl_typesupport_opensplice_cpp/resource/msg__type_support.cpp.em:
```diff
@@ -122,55 +142,57 @@ register_type__@(spec.base_type.type)(
-@[    elif field.type.is_primitive_type()]@
-      dds_message.@(field.name)_[i] = ros_message.@(field.name)[i];
+@[    if isinstance(member.type.basetype, BaseString)]@
+      dds_message.@(member.name)_[i] = ros_message.@(member.name)[i].c_str();
```
I am not sure if `c_str` is going to work for `WString`s.

jacobperron replied:
You're probably right, unless the ROS message is using a std::wstring. How is a WString represented in the ROS message and what should we do for the conversion?

dirk-thomas replied:
We don't have any examples / tests for that yet. The representation of WString in each language is described here: https://github.com/ros2/design/blob/gh-pages/articles/142_idl.md#type-mapping

jacobperron replied:
Ok, since `WString` is described as having type `std::u16string`, I think the use of `c_str()` should work since it is a [std::basic_string](https://en.cppreference.com/w/cpp/string/basic_string).

jacobperron committed:
Update typsupport_opensplice_cpp script to accept output path for converted IDL files

jacobperron committed:
Add function for converting from IDL to OpenSplice compatible IDL

jacobperron committed:
Minor fixups in cpp typesupport

jacobperron committed:
Update c typesupport message template

jacobperron committed:
Update C typesupport EmPy files

jacobperron committed:
Fixes to C++ typesupport

jacobperron committed:
Update C typesupport service templates

jacobperron committed:
Update C++ typesupport service templates

jacobperron committed:
Offload conversion stuff to rosidl_dds

jacobperron committed:
Rename generated external files in C ts

jacobperron committed:
Refactor rosidl_generator_dds_idl extension

jacobperron committed:
Fix C typesupport

jacobperron committed:
Fix C++ typesupport (messages compiling now!)

jacobperron committed:
Fixes post-rebase

jacobperron committed:
Fix template bug

jacobperron committed:
More fixes to C++ templates

jacobperron committed:
Fixes to templates (services compiling now)

jacobperron committed:
Pass in additional service template wrappers to rosidl_generator_dds_idl

jacobperron committed:
More fixes to templates

Moved header guards to top-level files.
Prepend type names to static variables.

jacobperron committed:
Add missing conversion function implementations and remove debug prints

jacobperron committed:
Call msg template from inside srv template

jacobperron committed:
Call msg template from inside srv template for C type support

jacobperron committed:
Fix lint

jacobperron commented:
CI with `idl-stage-7` branches, [ros2/rosidl (hidmic/idl-stage-7)](https://github.com/ros2/rosidl/tree/hidmic/idl-stage-7), and ros2/rosidl_dds#46 (jacob/idl-stage-7):
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5749)](http://ci.ros2.org/job/ci_linux/5749/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2333)](http://ci.ros2.org/job/ci_linux-aarch64/2333/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4750)](http://ci.ros2.org/job/ci_osx/4750/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5649)](http://ci.ros2.org/job/ci_windows/5649/)

dirk-thomas commented:
Once this passes CI please (squash) merge this into `idl-stage-7`.

jacobperron commented:
CI with `idl-stage-7` and  [ros2/rosidl (hidmic/idl-stage-7)](https://github.com/ros2/rosidl/tree/hidmic/idl-stage-7): 

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=5764)](http://ci.ros2.org/job/ci_linux/5764/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2349)](http://ci.ros2.org/job/ci_linux-aarch64/2349/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=4765)](http://ci.ros2.org/job/ci_osx/4765/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5666)](http://ci.ros2.org/job/ci_windows/5666/)

dirk-thomas commented:
The `idl-stage-7` branch can't compile anymore with this since the `rosidl_generator_dds_idl` interface is the same for Connext and OpenSplice typesupport. Therefore I am merging this...
